dips/dip-201.md

```diff
@@ -0,0 +1,79 @@
+---
+dip: 201
+title: 2-chain DiemBFT
+authors: Zekun Li (@zekun000)
+dip manager: Avery Ching (aching)
+status: Draft
+type: Standard
+created: 08/09/2020
+---
+
+# Summary
+
+This DIP describes a major latency improvement - 2-chain commit rule instead of 3-chain - brought by DiemBFT v4.
+
+# Abstract
+
+Currently, DiemBFT inherits the 3-chain commit rules from hotstuff for optimistic responsiveness. However, with timeout
+and round synchronization mechanism from DiemBFT, our research shows that we could get 2-chain commit rule almost for free.
+We only discuss engineering impact and omit the theoretic part from this DIP, readers are encouraged to read the full white paper.
+
+# Required Changes
+
+## Types
+In order to satisfy the safety requirement, two new timeout related structs are added to represent the highest quorum cert the validator node has when timeout.
+
+```rust
+pub struct TwoChainTimeout {
+    /// Epoch number corresponds to the set of validators that are active for this round.
+    epoch: u64,
+    /// The consensus protocol executes proposals (blocks) in rounds, which monotonically increase per epoch.
+    round: Round,
+    /// The highest quorum cert the signer has seen.
+    quorum_cert: QuorumCert,
+}
+pub struct TwoChainTimeoutCertificate {
+    timeout: TwoChainTimeout,
+    signatures: BTreeMap<Author, (Round, Ed25519Signature)>,
+}
+```
+
+TwoChainTimeoutCertificate only carries a single quorum cert instead of 2f+1 as an optimization to save both bandwidth and verification cost.
+It keeps the TwoChainTimeout with the highest quorum cert and the signature is signed on the derived struct
+```rust
+pub struct TimeoutSigningRepr {
+    pub epoch: u64,
+    pub round: Round,
+    pub hqc_round: Round,
+}
+```
+
+## Rules
+Safety Rules has been changed accordingly, proof of safety and liveness is in white paper.
+
+Voting Rule:
+1. block.round > last vote round
+2. block.round == block.qc.round + 1 || (block.round == tc.round + 1 && block.qc.round >= tc.highest_hqc_round)
+
+Timeout Rule:
+1. round >= last vote round
+2. (round == qc.round + 1 || round == tc.round + 1) && qc.round >= 1-chain round
+
+Commit Rule:
+ - It's safe to commit block A if there exist a 2-chain A <- QC <- B <- QC  such that b.round == a.round + 1.
+
+## Upgrade
+Switching from existing DiemBFT to the newer version requires reconfiguration. On-chain consensus config is added to support the transition.
+We also switch from BCS to a backward compatible wire protocol (JSON) to ease the complexity in migrating data structures.
+
+## Testing
+Consensus protocol change should be taken seriously since it's critical to the safety and liveness of the blockchain. Current plan includes
+1. Validate protocol and proof 
+2. Enable in all testing environment
+3. Add fault injection
+4. Stress test the reconfiguration switch
+5. Go through regular release pipeline
+
+## Client
+
+This change is client agnostic besides the latency improvement.
```

Comment:
```suggestion
This change is client agnostic, other than an observable improvement in commit latency improvement.
```

There is a slight nuance here that the LedgerInfo fields will reflect this change right?  For instance the rounds and such from BlockInfo will follow the 2-chain, but no one (e.g. clients) are supposed to be deriving consensus internals from such field.  Still might be useful to make sure that is clear here.

Reply:
agreed